processing/src/main/java/org/apache/druid/data/input/impl/AggregateProjectionSpec.java

```diff
@@ -0,0 +1,381 @@
+  }
+
+  @Override
+  public String toString()
+  {
+    return "AggregateProjectionSpec{" +
+           "name='" + name + '\'' +
+           ", groupingColumns=" + groupingColumns +
+           ", virtualColumns=" + virtualColumns +
+           ", aggregators=" + Arrays.toString(aggregators) +
+           ", ordering=" + ordering +
+           '}';
+  }
+
+  /**
+   * Ensure that the projection has the specified column required by a {@link CursorBuildSpec} in one form or another.
+   * If the column is a {@link VirtualColumn} on the build spec, ensure that the projection has an equivalent virtual
+   * column, or has the required inputs to compute the virtual column. If an equivalent virtual column exists, its
+   * name will be added to the 'rewriteVirtualColumns' parameter  so that the build spec name can be mapped to the
+   * projection name. If no equivalent virtual column exists, but the inputs are available on the projection to
+   * compute it, it will be added to 'referencedVirtualColumns' parameter.
+   *
+   * If the column is not a virtual column, {@link Projections.PhysicalColumnChecker} is a helper method which returns
+   * true if the column is present on the projection OR if the column is NOT present on the base table (meaning missing
+   * columns that do not exist anywhere do not disqualify a projection from being used).
+   */
+  private boolean hasRequiredColumn(
+      String column,
+      VirtualColumns buildSpecVirtualColumns,
+      Set<VirtualColumn> requiredBuildSpecVirtualColumns,
+      Map<String, String> mapBuildSpecColumnToProjectionColumn,
+      Projections.PhysicalColumnChecker physicalColumnChecker
+  )
+  {
+    final VirtualColumn buildSpecVirtualColumn = buildSpecVirtualColumns.getVirtualColumn(column);
+    if (buildSpecVirtualColumn != null) {
+      // check to see if we have an equivalent virtual column defined in the projection, if so we can
+      final VirtualColumn projectionEquivalent = virtualColumns.findEquivalent(buildSpecVirtualColumn);
+      if (projectionEquivalent != null) {
+        if (!buildSpecVirtualColumn.getOutputName().equals(projectionEquivalent.getOutputName())) {
+          mapBuildSpecColumnToProjectionColumn.put(
+              buildSpecVirtualColumn.getOutputName(),
+              projectionEquivalent.getOutputName()
+          );
+        }
+        return true;
+      }
+
+      requiredBuildSpecVirtualColumns.add(buildSpecVirtualColumn);
+      final List<String> requiredInputs = buildSpecVirtualColumn.requiredColumns();
+      if (requiredInputs.size() == 1 && ColumnHolder.TIME_COLUMN_NAME.equals(requiredInputs.get(0))) {
+        // wtb some sort of virtual column comparison function that can check if projection granularity time column
+        // satisifies query granularity virtual column
+        // can rebind? q.canRebind("__time", p)
+        // special handle time granularity
```

Comment:
yea sorry, still a bunch of todos and my rambling comments all over the place, this one is about wanting to dump using `Granularity` at all in favor of giving some way that a virtual column can decide if it can replace __time to check for things like finer granularity. i'm not going to do that in this PR, its just notes for myself, i'm still working on cleaning this up.